wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/hc/probe/ProbeScheduler.java

```diff
@@ -0,0 +1,203 @@
+   * since it is supposed that interruption is a very quick call. Separate thread is needed to
+   * prevent a situation when executor is full of jobs and current ones are hanging but we need to
+   * time them out.
+   */
+  private final Timer timeouts;
+  /** Mapping of workspaceId to a list of futures with probes of a workspace. */
+  private final Map<String, List<ScheduledFuture>> probesFutures;
+
+  @Inject
+  public ProbeScheduler(@Named("che.workspace.probe_pool_size") int probeSchedulerPoolSize) {
+    probesExecutor =
+        new ScheduledThreadPoolExecutor(
+            probeSchedulerPoolSize,
+            new ThreadFactoryBuilder().setDaemon(true).setNameFormat("ServerProbes-%s").build());
+    timeouts = new Timer("ServerProbesTimeouts", true);
+    probesFutures = new HashMap<>();
+  }
+
+  /**
+   * Schedules provided {@link WorkspaceProbes} and uses provided {@code Consumer<ProbeResult>} to
+   * send probes results. Respects {@link ProbeConfig} parameters such as thresholds, timeouts. Note
+   * that probe execution is not deleted automatically when probe results are passed to probe
+   * results consumer. To stop a probe execution method {@link #cancel(String)} should be used.
+   *
+   * @param probes probes to check
+   * @param probeResultConsumer consumer of {@link ProbeResult} instances produced on retrieving
+   *     probe execution results
+   */
+  public void schedule(WorkspaceProbes probes, Consumer<ProbeResult> probeResultConsumer) {
+    probesFutures.putIfAbsent(probes.getWorkspaceId(), new ArrayList<>());
+    probes
+        .getProbes()
+        .forEach(
+            probeFactory -> schedule(probes.getWorkspaceId(), probeFactory, probeResultConsumer));
+  }
+
+  /**
+   * Dismisses following and if possible current executions of probes of a workspace with a
+   * specified ID.
+   */
+  public void cancel(String workspaceId) {
+    List<ScheduledFuture> tasks = probesFutures.remove(workspaceId);
+    if (tasks == null) {
+      return;
+    }
+
+    tasks.forEach(task -> task.cancel(true));
+  }
+
+  private void schedule(
+      String workspaceId, ProbeFactory probeFactory, Consumer<ProbeResult> probeResultConsumer) {
+    ProbeConfig probeConfig = probeFactory.getProbeConfig();
+    Task task = new Task(probeFactory, probeResultConsumer);
+    // scheduleWithFixedDelay is used in favor of scheduleAtFixedRate because in case of big amount
+    // of scheduled probes start time of tasks may shift and this may lead to a situation when
+    // another probeConfig is needed immediately after the previous one is finished which doesn't
+    // seem a good thing
+    ScheduledFuture scheduledFuture =
+        probesExecutor.scheduleWithFixedDelay(
+            task,
+            probeConfig.getInitialDelaySeconds(),
+            probeConfig.getPeriodSeconds(),
+            TimeUnit.SECONDS);
+
+    List<ScheduledFuture> workspaceProbes =
+        probesFutures.computeIfPresent(
```

Comment:
since you are using `HashMap` you have no guarantee that `computeIfPresent` would be performed atomically, so potentially it can lead to non-deterministic behaviour.

Reply:
+1 thanks